### PR TITLE
Fix codegen for `codec::Compact` as type parameters

### PR DESCRIPTION
**Pull request by ascjones**
Fixes https://github.com/paritytech/cargo-contract/issues/734.

When a `codec::Compact` type appears as a type parameter e.g. `Option<Compact<u128>>` as it does in the contracts dispatchables e.g. https://github.com/paritytech/substrate/blob/4b7911607b48a5ca7ebf1d655f12d91cdd241866/frame/contracts/src/lib.rs#L498, then this will incorrectly result in a `Option<u128>` type being generated by `subxt-codegen`.

This is because the current codegen assumes that the compact value will be the type of the field directly and can generate the following:

`struct S { #[codec(compact)] field: u128 }`

The `compact` attribute will cause the conversion to be handled during encoding and decoding of the struct. This means that the "inner" type (in this case `u128`) can be used directly.

However when the compact type is a type parameter then it is required to use the `codec::Compact<T>` wrapper type directly instead of the field level attribute.

## Refactoring

This PR also incorporates some refactoring of the type path codegen, specifically expanding the `TypePathType` to a strongly typed enum which maps from the `scale_info::TypeDef` but with all the information necessary to generate the hierarchical type paths.

### CODE REVIEW

File: .rustfmt.toml [hunk 1]
```diff
@@ -9,7 +9,6 @@ format_code_in_doc_comments = false
 comment_width = 80
 normalize_comments = true # changed
 normalize_doc_attributes = false
-license_template_path = "FILE_TEMPLATE" # changed
```
Comment (jsdw) [inline .rustfmt.toml:12]: I never really noticed but does removing this mean no checking that the license header is at the top of files now?

File: codegen/src/types/tests.rs [hunk 2]
```diff
@@ -371,6 +371,43 @@ fn compact_fields() {
     )
 }
 
+#[test]
+fn compact_generic_parameter() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct S {
+        a: Option<<u128 as codec::HasCompact>::Type>,
+    }
```
Comment (jsdw) [inline codegen/src/types/tests.rs:380]: Is it worth adding tests to check that some of the other TypePath variants lead to the output you expect?

**Review by jsdw — APPROVED** [review]
Look good; tidy!

File: .rustfmt.toml (hunk 1 above)
Comment (ascjones) [reply .rustfmt.toml:12]: No it has been removed now so it doesn't do anything other than give a warning https://github.com/rust-lang/rustfmt/issues/5103

**Review by lexnv — APPROVED** [review]

**Pull request merged**